twa

```diff
@@ -391,8 +392,78 @@ function stage_4_scm_repos {
 # * Fetch the server's cert using `openssl s_client`, check it for known bad suites
 #   * Alternatively, just `nmap --script ssl-enum-ciphers -p 443 $domain`
 
+function stage_5_CAA {
+  verbose "Stage 5: Certification Authority Authorization"
+  verbose "Which Certificate Authorities (CAs) are allowed to issue certificates for the domain"
+    code=`dig +short caa $domain`
+    number_of_lines=`echo $code|wc -l`
+    if [ $number_of_lines -gt 1 ];then
+      INFO "multiple CAAs specified, everyone join the party"
+      #Cloudflare currently set the following set of certs as CAA
+      if [ echo $code|grep -c 'issue “comodoca.com”' -gt 0 ];then
+        if [ echo $code|grep -c 'issue “digicert.com”' -gt 0 ];then
+          if [ echo $code|grep -c 'issue “globalsign.com”' -gt 0 ];then
+            if [ echo $code|grep -c 'issuewild “comodoca.com”' -gt 0 ];then
+              if [ echo $code|grep -c 'issuewild “digicert.com”' -gt 0 ];then
+                if [ echo $code|grep -c 'issuewild “globalsign.com”' -gt 0 ];then
+                  INFO "CAA is probably provided by Cloudflare"
+                fi
+              fi 
+            fi
+          fi
+        fi
+      fi
+    fi
+
+
+    echo $code| while read line ; do
+      flag=`echo $code |cut -d' ' -f1| sed -e 's/^"//' -e 's/"$//'`
+      tag=`echo $code |cut -d' ' -f2`
+      value=`echo $code|cut -d' ' -f3| sed -e 's/^"//' -e 's/"$//'`
+      #list does not match returned names at this stage
+      #if [ `grep -c $value certs/untrustcert.txt` -ne 0 ];then    
+      #  FAIL "CA is listed as untrusted!"
+      #elif [ `grep -c $value certs/trustcert.txt` -ne 0 ];then        
+      #   PASS "CA is listed as trusted"
+      #else
+      #   UNK "Got an unknown CA:"$code
+      #fi
+
+      if [[ -z "$code" ]]; then
+         FAIL "No CAA record or doesn't specify an issuer at all. Potentially vunerable to have CA changed/MTM"
+      elif [[ ${code} == ";" ]]; then
+         MEH "CAA disallowed all CAA issuers"
+      elif [ `echo $tag| grep -c "issue"` -ne 1 ]; then
+         FAIL "CAA incorectly specified"
+      elif [ ${flag} -eq 0 ] && [ $tag == "issue" ]; then
+         PASS "CAA issuer:"$value
+      else
+         UNK "Got a weird response code:"$code
+      fi
+    done
+
+    nmap $domain|grep open|cut -d/ -f1 | while read line ; do
```

Comment:
This test is out-of-scope for a CAA check.

I'd like to add something like this in the future, but `nmap` is probably too chunky of a dependency. Ideally, we'd use `openssl` to enumerate a cipher list and then use `openssl s_client` to test those against the server.

Reply:
knowing what ports are open is important for both making sure the server is correctly secured and for detecting if it is running something like tomcat or websphere on 8080/8443. You have assumed everything runs on IIS/apache.

Reply:
> You have assumed everything runs on IIS/apache.

Well, not quite: for simplicity, `twa` just checks 80 and 443. That doesn't assume anything about the particular webserver being used.

I agree that testing a few common nonstandard ports is a good idea (8080 and 8443 definitely, but also 280, 8081, 4567, etc), but I think it's out of scope for this particular PR.